### PR TITLE
update code insee in seloger

### LINKED ISSUE #10: Code Postal vs Code Insee
toutes mes requetes de recherche manuel sont bases sur le [ci ](https://seloger.com/list.htm?tri=initial&enterprise=0&idtypebien=2,1&pxMax=2000000&idtt=2,5&naturebien=1,2,4&ci=750117)et non plus sur [cp ](https://github.com/ludaavics/pogam/blob/3e8bb4bb903a2de5e9182215337b4e7845e4560d/pogam/scrapers/seloger.py#L180) (code insee vs code postal), donc je narrive pas a trouver la bonne nomenclature pour 75016 (tried 75016, 75116, 75016,75116, 75011)

fix:
il faut introduire l idee que le code postal peut etre un cp ou un ci, et que si cp=75016, then use ci=750116

Autre point, j'ai verifie sur un bien dont je connais l'adresse, 'mapPoly' a l'air de donner un lieu plus precis pour le bien
Pour ce bien j'ai 110 rue de la pompe, alors que l'adresse est 178 av victor hugo, a cote mais pas exactement ca

### PR DESCRIPTION
**Pull request by chipsus**
Closes #10

### CODE REVIEW

**Comment by chipsus** [comment]
So I have tried that in couple of requests on my pc
to convert cp to ci I use an url, we might just store that csv on git

File: pogam/scrapers/seloger.py [hunk 1]
```diff
@@ -41,7 +41,18 @@
     )""",
     re.UNICODE | re.VERBOSE,
 )
-
+def MapCPtoCI (cp):
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: While there is nothing technically wrong with `CamelCaseFunctionNames`, the convention in python is to use `snake_case`, i.e. `map_cp_to_ci` in this case.

You can find the official python style [here](https://www.python.org/dev/peps/pep-0008/), in particular the part about naming functions [here](https://www.python.org/dev/peps/pep-0008/#id45). As a future reference, dor docstring [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html) is very popular for scientific application and the [Google style guide](https://github.com/google/styleguide/blob/gh-pages/pyguide.md) is ..well .. Google.

Generally speaking, when you contribute to a project you want to match the style of the project (even if it doesn't strictly adhere to the prescriptions in any of these guides).

File: pogam/scrapers/seloger.py [hunk 2]
```diff
@@ -41,7 +41,18 @@
...
+    try:
+        cp=str(cp)
+        df = pd.read_csv('https://www.data.gouv.fr/en/datasets/r/6d3428b2-3893-45a1-b404-2522a4e77d41',sep=';')
+        n=df[df['Code Postal'] == cp]['Code INSEE'].values[0]
+        n=str(n)
+        #get seloger readable ci with additional 0
+        n=n[0:2]+'0'+n[2:]
+        return n
+    except:
+        print('An error occurred converting code postal to code insee. Check for wrong postal code (75116 not recognised, use 75016)')
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: you never want to use `print` in production code. If you just want to display a message for information or debugging purposes, use the [logging](https://docs.python.org/3/library/logging.html) library. you should find plenty of examples in the pogam's code.

In this case, you actually want the program to stop! As it stands, since you caught the exception, your function will print, and since there is no `return`, it will implicitly return `None` and you will just end up with `None`'s in your list of postal_code. The code looks like it's running fine, but you're probably just not filtering for postcode

so you want to use raise an execption -- probably a `ValueError` in this case: `raise ValueError(f'Could not find insee code for postal code {ci}')` or something

File: pogam/scrapers/seloger.py [hunk 3]
```diff
@@ -41,7 +41,18 @@
...
+    """ convert CP to CI as seloger use CI by default for search."""
+    try:
+        cp=str(cp)
+        df = pd.read_csv('https://www.data.gouv.fr/en/datasets/r/6d3428b2-3893-45a1-b404-2522a4e77d41',sep=';')
+        n=df[df['Code Postal'] == cp]['Code INSEE'].values[0]
+        n=str(n)
+        #get seloger readable ci with additional 0
+        n=n[0:2]+'0'+n[2:]
+        return n
+    except:
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: This will  catch every single exception. For instance, if the file goes away, your `pd.read_csv` will error out, but will be interpret by your script as wrong post code input. It makes debugging more difficult (program is telling you "wrong post code", when the real error is missing file)

Best practices:
  - catch the specific exception(s) you anticipate
  - put as few lines of code as possible in the `try` block. 

In this case it looks like if the user supplies a wrong post code, `df[df['Code Postal'] == cp]['Code INSEE'].values` will be empty, so the `[0]` will raise an IndexError. So something like: 

```python
...
df = pd.read_csv(...)
try:
    n=df[df['Code Postal'] == cp]['Code INSEE'].values[0]
except IndexError:
    raise ValueError(f"Unknown post code f{cp}".)
n = str(n)
...
```

File: pogam/scrapers/seloger.py [hunk 4]
```diff
@@ -161,6 +172,11 @@ def seloger(
     min_beds = floor(min_beds) if min_beds is not None else min_beds
     max_beds = ceil(max_beds) if max_beds is not None else max_beds
 
+    # convert code postal to insee code
+    insee_codes=[]
+    for n in post_codes:
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: For very simple loops like these, list comprehensions are usually preferred, both for style and for performance reasons:

```python
insee_codes = [MapCPtoCI(cp) for cp in post_codes]
```
To be clear, for what we're doing here, what you have is perfectly fine.

Some references 
 -  https://leadsift.com/loop-map-list-comprehension/
 - https://towardsdatascience.com/how-list-comprehensions-can-help-your-code-look-better-and-run-smoother-3cf8f87172ae

**Review by ludaavics — CHANGES_REQUESTED** [review]
Great start, thanks for the pr!
A few minor things to change, but I think this is quite good and helpful.

I do think we might be able to improve the overall approach / strategy a bit: will follow up in a bit.

File: pogam/scrapers/seloger.py [hunk 5]
```diff
@@ -41,7 +43,20 @@
...
-
+def map_cp_to_ci (cp: str) -> str:
+    """ convert CP to CI as seloger use CI by default for search."""
+    df = pd.read_csv('https://www.data.gouv.fr/en/datasets/r/6d3428b2-3893-45a1-b404-2522a4e77d41',sep=';')
+    cp=str(cp)    
+    try:
+        n=df[df['Code Postal'] == cp]['Code INSEE'].values[0]
+    except Indexerror:
+        msg = f"Unknown post code f{cp}"
+        raise ValueError(msg)
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py:79]: `raise` will stop your program, so you will never reach the logger line.
you don't really need to log anything here, since you will see `msg` as an error msg. But if for some reason you wanted to, you should put it before the `raise`

**Comment by ludaavics** [comment]
Thanks for making the changes, @chipsus 

Taking a step back, i have two issues with the strategy in your PR.

First, i am reluctant to add the CSV file to the repo. As a general rule, you want to separate code from data as much as possible. In this particular case, googling around, it seems like INSEE codes are updated every year. I'm sure 99% of them don't change, but still -- it's a bit fragile to just commit the csv file directly.

One alternative would be to maintain a table of the mapping in our database. Doing some more research, however, it seems La Poste has a very nice [API](https://datanova.legroupe.laposte.fr/explore/dataset/laposte_hexasmal/api/?disjunctive.code_commune_insee&disjunctive.nom_de_la_commune&disjunctive.code_postal&disjunctive.libell_d_acheminement&disjunctive.ligne_5&refine.code_postal=92200)  that includes, among other things code postal / code insee mapping. For instance, this would be the request for poste code [92200](https://datanova.legroupe.laposte.fr/api/records/1.0/search/?dataset=laposte_hexasmal&facet=code_commune_insee&facet=nom_de_la_commune&facet=code_postal&facet=libell_d_acheminement&facet=ligne_5&refine.code_postal=92200). I think this is the best solution for post code / insee code mapping. We refer to an authority that's probably going to do a much better job of keeping things up to date than we ever could. This is what APIs are for. It _is_ a bit more work, but i think well worth it.

The other issue is this extra 0 business. This I am not sure how to deal with, to be honest. Both the csv data set and the La Poste API confirm that the correct insee code for 92200 should be 92051. Just adding an extra 0, like you did, doesn't work. If a search for post code 75009, i get ci=750109..

**Comment by ludaavics** [comment]
ok i think I might have something. 
It looks like seloger relies on an API for its auto complete of localities. In addition, it seems like the response of that API includes the post code. So we can:

 - hit that API to get the auto complete suggestions
 - scan through the list of suggestions, and look for one where the post code matches our desired post code.
 - throw an error if we can't find any.

You would just have to re-write your map cp_to_ci function. The snippet i used below when experimenting should get you started

```python
import requests  
import json
cp = 92200
url = f"https://autocomplete.svc.groupe-seloger.com/api/v2.0/auto/complete/fra/63/10/8/SeLoger?text={cp}"
response = requests.get(url)
cities = r.json()

# result:
[{'Type': 'Ville',
  'Display': '92200 Neuilly-sur-Seine',
  'Tag': 'Neuilly-sur-Seine',
  'Params': {'ci': '920051'},
  'Meta': {'Zips': ['92200'],
   'Quartiers': [{'Type': 'Quartier',
     'Id': '131762',
     'Name': 'Saint James-Général Koenig',
     'Poids': 403},
    {'Type': 'Quartier',
     'Id': '131765',
     'Name': "Ile de la Jatte-Parc d'Orléans",
     'Poids': 276},
    {'Type': 'Quartier',
     'Id': '131764',
     'Name': 'La Plaine des Sablons',
     'Poids': 205},
    {'Type': 'Quartier',
     'Id': '131763',
     'Name': 'Charles Laffitte',
     'Poids': 41},
    {'Type': 'Quartier',
     'Id': '131761',
     'Name': 'Saint James-Delabordère',
     'Poids': 17}],
   'HierarchicalZones': [{'Type': 'Departement',
     'Id': '17645',
     'Name': 'Hauts-de-Seine',
     'Poids': 21239},
    {'Type': 'Region', 'Id': '2238', 'Name': 'Île-de-France', 'Poids': 224200},
    {'Type': 'Pays', 'Id': '250', 'Name': 'France', 'Poids': 962515}]}}]
```

I am not sure what the `/fra/63/10/8/` part of the URL is, might have to play around a bit with that.

Think you can take a crack at it ?

File: pogam/scrapers/seloger.py [hunk 6]
```diff
@@ -41,7 +45,26 @@
...
+        a string of the corresponding code insee recognised by seloger
+    """
+    url = f"https://autocomplete.svc.groupe-seloger.com/api/v2.0/auto/complete/fra/63/10/8/SeLoger?text={cp}"
+    response = requests.get(url)
+    cities = response.json()
+    try:
+        ci=cities[0]['Params']['ci']
+    except Indexerror:
+        msg = f"Unknown post code f{cp}"
+        logger.debug(msg)
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: I don't think you need to log this message, because you will see it when your program raises the exception in the next line anyways.

File: pogam/scrapers/seloger.py [hunk 7]
```diff
@@ -41,7 +45,26 @@
     )""",
     re.UNICODE | re.VERBOSE,
 )
+def map_cp_to_ci (cp: str) -> str:
+    """
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: 👍Nice on the google-style docstrings

File: pogam/scrapers/seloger.py [hunk 8]
```diff
@@ -3,6 +3,10 @@
 import logging
 import random
 import re
+import pandas as pd
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: Since you no longer need `pd.read_csv`, you can remove this import

File: pogam/scrapers/seloger.py (hunk 7 above)
Comment (ludaavics) [reply pogam/scrapers/seloger.py]: As a side note, this is more than just 'nice to have'. The docstring in the code directly feed into the [documentations](https://ludaavics.github.io/pogam/api.html)

Not as important for this function, since it won't be exposed, but good to keep in mind.

File: pogam/scrapers/seloger.py [hunk 9]
```diff
@@ -41,7 +45,26 @@
     )""",
     re.UNICODE | re.VERBOSE,
 )
+def map_cp_to_ci (cp: str) -> str:
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: By convention, internal functions' name (functions that aren't meant to be called by the end-user) are prefixed with an underscore, so `_map_cp_to_ci`.
Or even better `_to_ci`, since the `cp` part is evident from the signature. 
Or even better-er, `_to_code_insee(code_postal)`: I try to avoid using shortcuts unless they are widely accepted ones. A little extra typing goes a long way in making your code more readable.

**Comment by ludaavics** [comment]
I've added some tests. 
Writing tests makes life easier once your codebase grows: they can help catch breaking changes in your code.

Writing tests also helps thinking through corner cases.

For instance,[here](https://github.com/chipsus/pogam/blob/patch-1/tests/test_seloger.py#L19) is a test that says "let's make sure that when i call `_to_code_insee` with post codes that i know to be wrong, such as, '-1' or 'Hello' or `None`, I actually get  a ValueErrors". 

But when i ran the test against our code (just type `make tests` in a console), the tests failed!  Indeed, if i feed 'hello' in our `_to_code_insee` function, we actually get an insee code back, instead of a value error we would want.

To understand why, remember we are using the auto complete api. So it accepts the name of the town (in addition to post codes). So when we type 'hello', the auto complete API returns a bunch of towns whose name look like 'hello'. Our code picks up the insee code of the first of those towns, a place called "Hellemelle"..

So once i saw the 'hello' tests were failing, i [changed the code](https://github.com/chipsus/pogam/commit/56e3b6d03bbb830126cc2ffeb94b21e9442020c7) to make sure we treated this case properly.  

* * *

I also tried to write test for departments. While trying to do that, I've found that se loger doesn't convert department codes, yielding, for instance, https://seloger.com/list.htm?tri=initial&enterprise=0&idtypebien=2,1&idtt=2,5&naturebien=1,2,4&cp=92 
If you try to submit both code postal and  department codes you will get both cp and ci: https://seloger.com/list.htm?tri=initial&enterprise=0&idtypebien=2,1&idtt=2,5&naturebien=1,2,4&cp=92&ci=750115

There is at least one notable exception of Paris (75) which returns a custom ci. 

I think we would want to support running a search on an entire department at once. So i think the logic should be

 -  convert everything we can to code insee
 - whatever we can't convert, assume it's a cp
 - submit a request with cp and / or ci

Do you want to take a crack at implementing that logic _and_ writing both positive and negative tests for it?

**Comment by chipsus** [comment]
I think this code does the job

I added a test to make sure we raise an error if we have any strings, wanted to have your thoughts versus your test to solve hello issue

How can I see the tests from make tests you have put in place ? 
I have tried make tests and I get an error that I can't understand
make: *** [Makefile:24: tests] Error 4

File: pogam/scrapers/seloger.py [hunk 10]
```diff
@@ -53,25 +54,36 @@ def _to_code_insee(post_code: str) -> str:
     Returns:
         Seloger's custom geographical codes; a modification of 'Code Insee'
     """
+    try: 
+        int(post_code)
+    except(ValueError):
+        msg = f"post code '{post_code}' is not an integer."
+        raise ValueError(msg)  
+
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: I am not sure i follow why we need this? 
There is an [exception today](https://www.google.com/search?q=lyon+metropoly+department+number), there were [exceptions in the past](https://en.wikipedia.org/wiki/Departments_of_France#Numbering) and there could be exceptions in the future. 
Not a huge deal, but might as well make the code robust to small arbitrary changes.

To be clear, arbitrary strings like the 'hello' example will be handled naturally [here](https://github.com/chipsus/pogam/blob/23da4f6551a479bc4a2c795bfbb3173183126990/pogam/scrapers/seloger.py#L75-L77): in the autocomplete results there just won't be any result with post code 'hello'.

File: pogam/scrapers/seloger.py [hunk 11]
```diff
@@ -53,25 +54,36 @@ def _to_code_insee(post_code: str) -> str:
...
-    cities = response.json()
-    city = [
-        city for city in cities if post_code in city.get("Meta", {}).get("Zips", [])
-    ]
-    if len(city) != 1:
-        msg = f"Unknown post code '{post_code}'."
-        raise ValueError(msg)
+    
+    if int(post_code) in range(1,96):
+        code_insee=post_code
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: Again, not really wrong, but kinda fragile, and doesn't work for DOM-TOM's or, more importantly, for Ile de france (75) where seloger returns `ci=750056`

File: pogam/scrapers/seloger.py [hunk 12]
```diff
@@ -213,7 +224,7 @@ def seloger(
     params: Dict[str, Union[float, str]] = {
         "projects": transaction,
         "types": ",".join(map(str, property_types)),
-        "places": "[" + "|".join([f"{{ci:{ic}}}" for ic in insee_codes]) + "]",
+        "places": "[" + "|".join([f"{{ci:{ic}}}" for ic in insee_codes]) + "]"+"[" + "|".join([f"{{cp:{dp}}}" for dp in department_codes]) + "]",
```
Comment (ludaavics) [inline pogam/scrapers/seloger.py]: 👍

File: pogam/scrapers/seloger.py (hunk 11 above)
Comment (ludaavics) [reply pogam/scrapers/seloger.py]: Actually -- i stand corrected on the above, re: ile de france. Looks like both cp=75 and ci=750056 work.

**Review by ludaavics — DISMISSED** [review]
Thanks for the update!
Your implementation definitely works; i've just tweaked a few things to, hopefully, make things a bit less fragile (any scraping is inherently very fragile).

Otherwise, looks good to me!
